v0.1 base64converter

lokihi opened the pull request.

Commit:
v0.1 base64converter

aasirotkin commented on CMakeLists.txt:
```diff
@@ -0,0 +1,41 @@
+enable_testing()
+
+add_executable(
+  base64_basic_test
```
Зачем два executable? одного будет достаточно

lokihi replied:
Сделал

aasirotkin commented on README.md:
```diff
@@ -0,0 +1,3 @@
+Basic header-only cpp library for base64:
+- Encoding
```
В readme хорошо ещё писать зависимости и стандарт языка. Ещё примеры использования можно добавить.

aasirotkin replied:
Добавь пожалуйста и расскажи, что можно декодировать двумя методами, а какими прочитаешь внизу

lokihi replied:
Сделал

aasirotkin commented on include/base64.hpp:
```diff
@@ -0,0 +1,201 @@
+{
+
+// Encoding table for base64 converter //
+std::array<char, 64> constexpr encodeTable
```
constexpr должно быть в самом начале и везде ниже тоже

lokihi replied:
Сделал

aasirotkin commented on include/base64.hpp:
```diff
@@ -0,0 +1,201 @@
+    encodedString.reserve(sizeEncoded);
+
+    // Storing pointer to currently encoded symbol in input string //
+    const uint8_t *inputIterator = reinterpret_cast<const uint8_t *>(&inputString[0]);
```
Тут const перед, а сверху constexpr после? Должно быть везде однотипно, как здесь

lokihi replied:
Сделал

aasirotkin commented on include/base64.hpp:
```diff
@@ -0,0 +1,201 @@
+inline std::string encode(std::string_view inputString)
+{
+
+    // Size of input string //
```
Не рекомендуется писать очевидные комментарии. Лучше тут и ниже везде убрать, где очевидно, что делается

lokihi replied:
Сделал

aasirotkin commented on include/base64.hpp:
```diff
@@ -0,0 +1,201 @@
+    // Iterate over blocks of 3 until the end of the input string //
+    for (size_t i = inputDataSize / 3; i; --i)
+    {
+
```
Лишний перенос строки

lokihi replied:
Сделал

aasirotkin commented on include/base64.hpp:
```diff
@@ -0,0 +1,201 @@
+    const uint8_t *inputIterator = reinterpret_cast<const uint8_t *>(&inputString[0]);
+
+    // Iterate over blocks of 3 until the end of the input string //
+    for (size_t i = inputDataSize / 3; i; --i)
```
```
for (size_t i = inputDataSize / 3; i; --i)
```
какое-то хитрое условие выхода. Перепиши пожалуйста так, чтобы было очевидно

lokihi replied:
Сделал

aasirotkin commented on include/base64.hpp:
```diff
@@ -0,0 +1,201 @@
+
+        std::uint32_t const concatenatedBits = (symbol1 << 16) | (symbol2 << 8) | symbol3;
+
+        encodedString.push_back(encodeTable[(concatenatedBits >> 18) & 0b0011'1111]);
```
Весь это цикл (for) рекомендую вынести в отдельный метод. Подумай сам как и что он должен делать. Расскажи мне прежде чем писать.

lokihi replied:
По поводу отдельного метода, можно тогда в функцию вынести для кодирования блока из 3 символов
Или если весь цикл for выносить, то можно тогда string_view передавать,  а возвращать строку без добавления padding char в конце

aasirotkin replied:
думаю, что весь for хорошо будет вынести

aasirotkin replied:
и думаю что ты правильно понял идею. Свитч скорее всего тоже можно вынести))

aasirotkin replied:
но в другой метод

lokihi replied:
А можно передавать в метод указатель на уже созданную строку для вывода, и сделать функцию void?

aasirotkin replied:
давай так попробуем. Сделай это в namespace detail {} внутри твоего namespace

lokihi replied:
Сделал

aasirotkin commented on include/base64.hpp:
```diff
@@ -0,0 +1,201 @@
+
+inline std::string decode(std::string_view inputString)
+{
+
```
Лишний перенос

lokihi replied:
Сделал

aasirotkin commented on include/base64.hpp:
```diff
@@ -0,0 +1,201 @@
+    const size_t inputDataSize = inputString.size();
+
+    // Ammount of paddings in input string //
+    const size_t numPadding = std::count(inputString.rbegin(), inputString.rbegin() + 4, '=');
```
Насколько эффективно искать эти символы в 3 циклах?
Сейчас оставь так, а потом может быть сделаем за один цикл

lokihi replied:
Циклы для поиска Return и Newline лишними были. Пытался одно исключение проверить и забыл удалить.  В цикл для декодирования уже встроено игнорирование этих символов

lokihi replied:
Сделал

aasirotkin commented on include/base64.hpp:
```diff
@@ -0,0 +1,201 @@
+    // Check, if ammount of padding symbols is correct (< 2) //
+    if (numPadding > 2)
+    {
+        throw std::runtime_error{
```
Сейчас оставь так, а потом переделаем это место, чтобы не было исключений

aasirotkin commented on include/base64.hpp:
```diff
@@ -0,0 +1,201 @@
+    }
+
+    // Calculate and reserve place for decoded string //
+    size_t sizeDecoded = (inputDataSize / 4 * 3) - numPadding;
```
Почему не const?

lokihi replied:
Сделал

aasirotkin commented on include/base64.hpp:
```diff
@@ -0,0 +1,201 @@
+    const uint8_t *inputIterator = reinterpret_cast<const uint8_t *>(&inputString[0]);    
+
+    uint8_t currentBlock[4];
+    int counterBlock = 0;
```
Почему int? везде size_t

lokihi replied:
Сделал

aasirotkin commented on include/base64.hpp:
```diff
@@ -0,0 +1,201 @@
+    int counterBlock = 0;
+
+    // Iterate over blocks of 4 until the end of the input string //
+    for (size_t i = 0; i < inputDataSize - (numPadding!=0)*4; i++)
```
Посчитай размер заранее inputDataSize - (numPadding!=0)*4;

lokihi replied:
Сделал

aasirotkin commented on include/base64.hpp:
```diff
@@ -0,0 +1,201 @@
+        std::uint32_t const concatenatedBits =
+            (decodeTable[currentBlock[0]] << 18) | (decodeTable[currentBlock[1]] << 12) | (decodeTable[currentBlock[2]] << 6) | decodeTable[currentBlock[3]];
+        decodedString.push_back((concatenatedBits >> 16) & 0b1111'1111);
+        decodedString.push_back((concatenatedBits >> 8) & 0b1111'1111);
```
Подумай как декомпозировать эту функцию тоже. Она очень длинная, сложно читать

lokihi replied:
Сделал

aasirotkin commented on include/base64.hpp:
```diff
@@ -0,0 +1,201 @@
+    }
+
+    return decodedString;
+}
```
Нужен перенос строки

lokihi replied:
Сделал

aasirotkin commented on test/base64_file_test.cpp:
```diff
@@ -0,0 +1,47 @@
+    std::ifstream t("resources/test.exe");
+    std::stringstream buffer;
+    buffer << t.rdbuf();
+    std::string const expected{buffer.str()};
```
ты уверен, что строка не пустая?

lokihi replied:
Сделал

aasirotkin commented on test/base64_test.cpp:
```diff
@@ -0,0 +1,100 @@
+#include "../include/base64.hpp"
+#include <string>
+
+// TEST FOR EMPTY INPUT //
```
Лишние комментарии, они дублируют названия тестов

lokihi replied:
Сделал

aasirotkin reviewed (approved):
Написал замечания, если что непонятно, пиши

aasirotkin commented on test/base64_file_test.cpp:
```diff
@@ -0,0 +1,47 @@
+    std::string const expected{buffer.str()};
+    std::string const actual{base64::decode({base64::encode({buffer.str()})})};
+    ASSERT_EQ(expected, actual);
+}  
```
Обрати внимание, у некоторых файлов есть такие кирпичи в конце. Это значит, что нет переноса строки. Рекомендую всегда делать перенос на новую строку.

lokihi replied:
Сделал

Commit:
Updated CMake file to use only one executable

Commit:
Decompose of encode and decode function

Commit:
Deleted unnecessary repeated comments

Commit:
Fixed tests paths to test files

aasirotkin commented on include/base64.hpp:
```diff
@@ -0,0 +1,207 @@
+        outputString.push_back(encodeTable[(symbol2 << 2) & 0b0011'1111]);
+        outputString.push_back('=');
+        break;
+    }
```
Некоторые компиляторы могут думать, что ты забыл default и будут делать предупреждение, а у нас 0-ая политика по отношению к предупреждениям, то есть это будет ошибка. Добавь пожалуйста default

lokihi replied:
Сделал

aasirotkin commented on include/base64.hpp:
```diff
@@ -0,0 +1,207 @@
+    for (size_t i = 0; i < blocksAmmount; i++)
+    {   
+        currentBlock[counterBlock] = *inputIterator++;
+        if (currentBlock[counterBlock] != '\n' && currentBlock[counterBlock]!='\r') counterBlock++;
```
По кодстайлу так нельзя, нужно перенести тело if на новую строку

lokihi replied:
Сделал

aasirotkin commented on include/base64.hpp:
```diff
@@ -0,0 +1,207 @@
+        if (currentBlock[counterBlock] != '\n' && currentBlock[counterBlock]!='\r') counterBlock++;
+        if (counterBlock == 4)
+        {
+        std::uint32_t const concatenatedBits =
```
Тут поехало форматирование

lokihi replied:
Сделал

aasirotkin commented on include/base64.hpp:
```diff
@@ -0,0 +1,207 @@
+        if (counterBlock == 4)
+        {
+        std::uint32_t const concatenatedBits =
+            (decodeTable[currentBlock[0]] << 18) | (decodeTable[currentBlock[1]] << 12) | (decodeTable[currentBlock[2]] << 6) | decodeTable[currentBlock[3]];
```
Форматируй так, чтобы было меньше 120 символов

lokihi replied:
Сделал

aasirotkin commented on include/base64.hpp:
```diff
@@ -0,0 +1,207 @@
+        break;
+    }
+    }
+}
```
Нужен перенос строки

lokihi replied:
Сделал

aasirotkin commented on include/base64.hpp:
```diff
@@ -0,0 +1,207 @@
+    const size_t sizeEncoded = (inputDataSize + 2) / 3 * 4;
+
+    std::string encodedString;
+    encodedString.reserve(sizeEncoded);
```
Может reserve перенесём в mainPartString? а то там push back делается, и не очевидно, что ранее был reserve

lokihi replied:
Сделал

aasirotkin commented on include/base64.hpp:
```diff
@@ -0,0 +1,207 @@
+    return encodedString;
+}
+
+
```
Лишние переносы строк

lokihi replied:
Сделал

aasirotkin commented on include/base64.hpp:
```diff
@@ -0,0 +1,207 @@
+    const size_t sizeDecoded = (inputDataSize / 4 * 3) - numPadding;
+
+    std::string decodedString;
+    decodedString.reserve(sizeDecoded);
```
Возможно тоже самое, посмотри

lokihi replied:
Сделал

aasirotkin commented on include/base64.hpp:
```diff
@@ -0,0 +1,207 @@
+    if (numPadding > 2)
+    {
+        throw std::runtime_error{
+            "Invalid base64 string. Couldn't resolve ammount of paddinngs."};
```
Давай сделаем так. Добавим ещё одну функцию decode(inputString, bool isError).

Если пользователь вызывает без флага, то выбрасывается исключение, а если с флагом, то флаг выставляется в false и возвращается пустая строка

aasirotkin replied:
И ещё, забыл, напиши на этот случай тесты, и с флагом и с исключением

lokihi replied:
То есть это как доп параметр в функцию подаётся?

aasirotkin replied:
У нас будет две функции
decode(std::string_view inputString) - тут выбрасывается исключение
decode(std::string_view inputString, bool& isOk) - тут isOk = false, если что не так пошло

Так в бусте делают, только у них там error_code.

lokihi replied:
Сделал

aasirotkin commented on include/base64.hpp:
```diff
@@ -0,0 +1,207 @@
+{
+    if (inputString.empty())
+    {
+        return std::string{};
```
Может давай уберём эту проверку? Насколько это частый кейс, что мы конвертим пустую строку? А вид портит

aasirotkin replied:
Тебе нужно, чтобы строка была не пустая, а может быть кратная 4?

lokihi replied:
Проверку на кратность с учетом newline добавил. Конвертацию пустой строки я бы оставил, так как иначе segfault можно словить при определении количества padding char в конце строки

lokihi replied:
Сделал

aasirotkin commented on test/base64_test.cpp:
```diff
@@ -0,0 +1,88 @@
+
+TEST(base64encode, InputEmpty)
+{
+    std::string const expected{};
```
В тестах везде тоже нужно поправить и перенести const в начало

lokihi replied:
Сделал

aasirotkin commented on test/base64_test.cpp:
```diff
@@ -0,0 +1,88 @@
+#include <gtest/gtest.h>
+
+#include "../include/base64.hpp"
```
по кодстайлу сначала подключаем сторонние файлы, потом, через перенос строки, свои

lokihi replied:
Сделал

aasirotkin commented on include/base64.hpp:
```diff
@@ -0,0 +1,207 @@
+    '4', '5', '6', '7', '8', '9', '+', '/'
+};
+
+// Decoding table for base 64 converter //
```
Тоже вот, как будто бы это очевидно из названии переменной

lokihi replied:
Сделал

aasirotkin commented on include/base64.hpp:
```diff
@@ -0,0 +1,207 @@
+{
+    const uint8_t *inputIterator = reinterpret_cast<const uint8_t *>(&inputString[0]);
+
+    for (size_t i = 0; i < inputString.size() / 3; i++)
```
Погугли чем отличается i++ и ++i, выбери тут подходящий

lokihi replied:
Сделал

aasirotkin commented on include/base64.hpp:
```diff
@@ -0,0 +1,207 @@
+    {
+        const uint8_t symbol1 = *inputIterator++;
+        const uint8_t symbol2 = *inputIterator++;
+        const uint8_t symbol3 = *inputIterator++;
```
Тоже самое, чем отличается ++ после, от ++ до

lokihi replied:
Сделал

aasirotkin commented on include/base64.hpp:
```diff
@@ -0,0 +1,207 @@
+{
+    switch (inputString.size() % 3)
+    {
+    case 0:
```
Вот тут можно сделать комментарий
case 0: // Что значит этот ноль
case 1: // Что такое один
и так далее

lokihi replied:
Сделал

aasirotkin commented on include/base64.hpp:
```diff
@@ -0,0 +1,207 @@
+    decodedString.reserve(sizeDecoded);
+
+    detail::decodingMainPartString(inputString, numPadding, decodedString);
+    detail::decodingPaddingChar(inputString, numPadding, decodedString);
```
Сверху, в encode, есть перенос между вызовами функций, а тут нет

aasirotkin reviewed (approved):
Ответил. Давай поправим эти замечания и думаю выйдем на финишную прямую.
Если сделал какой-то пункт, то напиши в сообщении "Сделал", я тогда будут понимать, что этот вопрос можно закрыть.

Commit:
Code style changes

aasirotkin commented on test/base64_file_test.cpp:
```diff
@@ -0,0 +1,46 @@
+    buffer << t.rdbuf();
+    const std::string expected{buffer.str()};
+    const std::string actual{base64::decode({base64::encode({buffer.str()})})};
+    ASSERT_EQ(expected, actual);
```
Добавь проверку, что expected не пустая

lokihi replied:
Сделал

aasirotkin reviewed (changes requested):
Дим, не делай resolve, это я должен сделать. Просто как-то маякни, что всё готово, напиши сделал или ок или ещё что, можно плюсик поставить)

aasirotkin commented:
И там сверху у тебя ещё есть не выполненные пункты

Commit:
Empty string check is added

Commit:
Tests for decode function with flag added

Commit:
New decode function and code styles fixes

- New decode function with error flag added
- Unnecessary comments delted
- Default statement for switches added
- Incrementation changed from i++ to ++i

lokihi committed:
Update README.md

lokihi committed:
Update README.md

aasirotkin commented on README.md:
```diff
@@ -0,0 +1,39 @@
+
+* Decoding with flag variable
+
+If an exception is caught, the value of the иoolean variable will be changed to false
```
Опечатка

lokihi replied:
Исправил

aasirotkin commented on include/base64.hpp:
```diff
@@ -0,0 +1,251 @@
+
+    for (size_t i = 0; i < inputString.size() / 3; ++i)
+    {
+        const uint8_t symbol1 = *inputIterator++;
```
у итератора ты решил не менять?

lokihi replied:
Как я понял, в цикле такое такое инкрементирование эффективнее, а здесь нам нужно сначала значение получать, а потом обращаться к следующему элементу строки

aasirotkin commented on include/base64.hpp:
```diff
@@ -0,0 +1,251 @@
+    {   
+        currentBlock[counterBlock] = *inputIterator++;
+        if (currentBlock[counterBlock] != '\n' && currentBlock[counterBlock]!='\r') 
+        counterBlock++;
```
Почему тут ++ после?

aasirotkin replied:
и форматирование поехало

aasirotkin commented on include/base64.hpp:
```diff
@@ -0,0 +1,251 @@
+
+    for (size_t i = 0; i < blocksAmmount; ++i)
+    {   
+        currentBlock[counterBlock] = *inputIterator++;
```
Почему тут ++ после?

aasirotkin commented on include/base64.hpp:
```diff
@@ -0,0 +1,251 @@
+    }
+}
+
+inline void decodingMainPartString(std::string_view inputString, int numPadding, std::string &outputString)
```
Почему numPadding имеет тип int?

lokihi replied:
Исправил

aasirotkin commented on include/base64.hpp:
```diff
@@ -0,0 +1,251 @@
+        counterBlock++;
+        if (counterBlock == 4)
+        {
+        std::uint32_t const concatenatedBits = (decodeTable[currentBlock[0]] << 18) | (decodeTable[currentBlock[1]] << 12) |
```
отступ поехал

aasirotkin replied:
и почему const посередине?

lokihi replied:
Исправил

aasirotkin commented on include/base64.hpp:
```diff
@@ -0,0 +1,251 @@
+    }
+}
+
+inline void decodingPaddingChar(std::string_view inputString, int numPadding, std::string &outputString)
```
почему numPadding int? он может быть отрицательным?

lokihi replied:
Исправил

aasirotkin commented on include/base64.hpp:
```diff
@@ -0,0 +1,251 @@
+{
+    switch (numPadding)
+    {
+    case 0:
```
нужен комментарий к 0 1 2

lokihi replied:
Сделал

aasirotkin commented on include/base64.hpp:
```diff
@@ -0,0 +1,251 @@
+    {
+        break;
+    }
+    
```
лишний перенос

lokihi replied:
Исправил

aasirotkin commented on test/base64_file_test.cpp:
```diff
@@ -0,0 +1,50 @@
+#include "../include/base64.hpp"
+
+#include <gtest/gtest.h>
```
Долгая история, но лучше отделять разные библиотеки при подключении. То есть тут есть gtest и std.
Добавь перенос

aasirotkin replied:
и так везде, не только здесь

lokihi replied:
Исправил

aasirotkin commented on test/base64_test.cpp:
```diff
@@ -0,0 +1,120 @@
+{
+    const bool expected{false};
+    bool flag;
+    const std::string decoded{base64::decode({"123"},flag)};
```
Пробел после запятой

lokihi replied:
Исправил

aasirotkin commented on test/base64_test.cpp:
```diff
@@ -0,0 +1,120 @@
+    const bool expected{false};
+    bool flag;
+    const std::string decoded{base64::decode({"123"},flag)};
+    ASSERT_EQ(expected, flag);
```
Тут лучше писать ASSERT_TRUE или ASSERT_FALSE, тогда не нужна будет переменная expected.

lokihi replied:
Исправил

aasirotkin commented on include/base64.hpp:
```diff
@@ -0,0 +1,251 @@
+    const size_t inputDataSize = inputString.size();
+
+    if ((inputDataSize-ammountNewline)%4!=0)
+        isOk = false;
```
Может быть не нужно идти дальше, если isOk уже false?

lokihi replied:
Исправил

aasirotkin commented on include/base64.hpp:
```diff
@@ -0,0 +1,251 @@
+    const size_t ammountNewline = std::count(inputString.begin(), inputString.end(), '\n');    
+    const size_t inputDataSize = inputString.size();
+
+    if ((inputDataSize-ammountNewline)%4!=0)
```
!= нужно обособить пробелами

lokihi replied:
Исправил

aasirotkin commented on include/base64.hpp:
```diff
@@ -0,0 +1,251 @@
+    return encodedString;
+}
+
+inline std::string decode(std::string_view inputString)
```
Самое главное.

У тебя получилось два одинаковых метода. Преимущество этого метода в том, что у тебя сообщение есть, но в версии с bool такого сообщения нет. Предлагаю тут тоже отказаться от двух разных сообщений.

Тогда тебе нужно сделать следующее.

```
inline std::string decode(std::string_view inputString)
{
    bool isOk = false;
    auto result = decode(inputString, isOk);
    if (!isOk)
        throw exception here;
    return result;
}
```

lokihi replied:
Сделал

Commit:
Updated Readme

Commit:
Code style changes

Commit:
Decode function fixes and code style changes

aasirotkin commented on include/base64.hpp:
```diff
@@ -0,0 +1,233 @@
+    const size_t numPadding = std::count(inputString.rbegin(), inputString.rbegin() + 4, '=');
+
+    if (numPadding > 2)
+        isOk = false;
```
Ну а тут почему не вернул сразу строку?

lokihi replied:
Сделал

Commit:
Added return of empty string

aasirotkin commented on include/base64.hpp:
```diff
@@ -0,0 +1,236 @@
+    for (size_t i = 0; i < blocksAmmount; ++i)
+    {   
+        currentBlock[counterBlock] = *inputIterator++;
+        if (currentBlock[counterBlock] != '\n' && currentBlock[counterBlock]!='\r') 
```
Пробельчики вокруг !=

lokihi replied:
Сделал

aasirotkin commented on include/base64.hpp:
```diff
@@ -0,0 +1,236 @@
+
+inline std::string decode(std::string_view inputString, bool& isOk)
+{
+    isOk = true;
```
Дим, смотри. Из-за этого true у тебя получается 6 дополнительных строк, 4 - на скобочки и 2 на isOk = false.
А если ты тут сделаешь isOk = false, то скобочки уходят и ты просто возвращаешь string{};, кстати скобочки можно убрать и у проверки на пустую строку, кодстайл это позволяет.

Итого, меняем true на false, и в самом конце не забываем выставить true, вместо 6 дополнительных строк, получаем 1 дополнительную. Плюс такого подхода ещё и в том, что не нужно выставлять false посередине, не нужно про это помнить.

Вот такой подход, очень поход на RAII, если не знаешь, то почитай

lokihi replied:
Сделал

aasirotkin commented on include/base64.hpp:
```diff
@@ -0,0 +1,236 @@
+    return decodedString;
+}
+
+} // namespace base64
```
Опять этот Тупик тут) Рекомендую добавить перенос строки

lokihi replied:
Сделал

aasirotkin commented on test/base64_test.cpp:
```diff
@@ -0,0 +1,117 @@
+TEST(base64decode, FlagCorrectNumPadding)
+{
+    bool flag;
+    const std::string decoded{base64::decode({"123456=="},flag)};
```
Где-то есть пробел после запятой, а где-то нет.
Добавь везде пожалуйста.

Однотипность оформления однажды очень тебе поможет. К примеру когда будешь писать regexp какой-нибудь

lokihi replied:
Исправил

aasirotkin reviewed (approved):
Дима, это точно последнее! До обеда сделаем и попрощаемся)))

Commit:
Minor changes in decode function and code style

Commit:
Minor changes to code style
